### PR TITLE
FIX: Fetch jQuery from a CDN since it can't come from admin

### PR DESCRIPTION
**Pull request by sunnysideup**
Sorry for the linting cleanup.  Do with the request as you see fit.

### CODE REVIEW

**Comment by GuySartorelli** [comment]
Looks like this is to fix https://github.com/symbiote/silverstripe-multivaluefield/issues/89 - thanks for this!

File: src/Fields/KeyValueField.php [hunk 1]
```diff
@@ -81,13 +81,13 @@ public function __construct($name, $title = null, $sourceKeys = [], $sourceValue
     public function Field($properties = [])
     {
         if (Controller::curr() instanceof ContentController) {
-            Requirements::javascript('silverstripe/admin: thirdparty/jquery/jquery.js');
+            Requirements::javascript('https://code.jquery.com/jquery-3.6.3.min.js');
```
Comment (GuySartorelli) [inline src/Fields/KeyValueField.php:87]:
```suggestion
            Requirements::javascript('//code.jquery.com/jquery-3.6.3.min.js');
```
I'm unsure if `Requirements::javascript()` accepts `//` or not.... if it doesn't, https will do.

File: src/Fields/MultiValueDropdownField.php [hunk 2]
```diff
@@ -46,21 +46,21 @@ public function setSource(array $source)
     public function Field($properties = [])
     {
         if (Controller::curr() instanceof ContentController) {
-            Requirements::javascript('silverstripe/admin: thirdparty/jquery/jquery.js');
+            Requirements::javascript('https://code.jquery.com/jquery-3.6.3.min.js');
```
Comment (GuySartorelli) [inline src/Fields/MultiValueDropdownField.php:52]:
```suggestion
            Requirements::javascript('//code.jquery.com/jquery-3.6.3.min.js');
```

File: src/Fields/KeyValueField.php (hunk 1 above)
Comment (GuySartorelli) [reply src/Fields/KeyValueField.php:84]: I'm also unsure if `Requirements::javascript()` accepts `//` or not.... if it doesn't, https will do.

File: src/Fields/MultiValueListField.php [hunk 3]
```diff
@@ -28,12 +28,12 @@ public function __construct($name, $title = null, $source = [], $value = null)
     public function Field($properties = [])
     {
         if (Controller::curr() instanceof ContentController) {
-            Requirements::javascript('silverstripe/admin: thirdparty/jquery/jquery.js');
+            Requirements::javascript('https://code.jquery.com/jquery-3.6.3.min.js');
```
Comment (GuySartorelli) [inline src/Fields/MultiValueListField.php:34]:
```suggestion
            Requirements::javascript('//code.jquery.com/jquery-3.6.3.min.js');
```

File: src/Fields/MultiValueTextField.php [hunk 4]
```diff
@@ -25,17 +25,17 @@ class MultiValueTextField extends FormField
     public function Field($properties = [])
     {
         if (Controller::curr() instanceof ContentController) {
-            Requirements::javascript('silverstripe/admin: thirdparty/jquery/jquery.js');
+            Requirements::javascript('https://code.jquery.com/jquery-3.6.3.min.js');
```
Comment (GuySartorelli) [inline src/Fields/MultiValueTextField.php:37]:
```suggestion
            Requirements::javascript('//code.jquery.com/jquery-3.6.3.min.js');
```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
This will cause problems for anyone who is using `Requirements::block('silverstripe/admin: thirdparty/jquery/jquery.js')` to block this js on the frontend (e.g. if they're using their own jquery already)... which I should have thought of when I did the change in silverstripe/admin, oops.

Anyway, can you please also add a configuration option to optionally _not_ output this file? Probably put it on `MultiValueTextField` since that's the superclass for _most_ of these.

**MultiValueTextField.php**
```php
private static bool $output_jquery_on_frontend = true;
```

**In each class**
```php
if (MultiValueTextField::config()->get('output_jquery_on_frontend')) {
  Requirements::javascript('https//code.jquery.com/jquery-3.6.3.min.js');
}
```

Please also create a PR for [`silverstripe/developer-docs`](https://github.com/silverstripe/developer-docs) which adds a note about this change to the `4.13.0` changelog, since this is a supported module.

**Comment by sunnysideup** [comment]
from what I know https:// is better than // - but I can remember exactly why.

**Comment by sunnysideup** [comment]
I was wondering if there should be a silverstripe-include-jquery module to have one unified strategy for including jquery. Even though, in a perfect world, jquery would be gone, lots of modules use it, etc... removing it from admin has created a lot of issues all over the show, and so another strategy might be to simply bring it back to admin as a sort of "that is just where it lives" approach. 

e.g. I was wondering what CDN would be best - and why , etc... 

Having a simple module that can be included to have jquey loaded, and only once, with automatic patches to its versions, as required, may be useful?

**Comment by GuySartorelli** [comment]
> removing it from admin has created a lot of issues all over the show, and so another strategy might be to simply bring it back to admin as a sort of "that is just where it lives" approach.

It's honestly a surprise to me that so many modules were referencing it - it was well known to be a very old version of jQuery and we had been under the assumption that it was _only_ used in the CMS... the fact that so many places have chosen to use it for front-end uses is surprising to me. I'd be very hesitant to put even the new version of jQuery there, since realistically admin javascript is for the CMS and using it in the frontend is an anti-pattern, generally speaking.

> Having a simple module that can be included to have jquey loaded, and only once, with automatic patches to its versions, as required, may be useful?

I'd honestly suggest it'd be better to rewrite modules like this one to use pure javascript instead, and therefore completely remove the reliance on jQuery. But either approach is out of scope for this PR, I'd say.

**Comment by sunnysideup** [comment]
Agreed with all that. Of course pure JS is better, but at times, you have a great piece of functionality that relies on jQuery plugin or something like that. In that case, it would be good to say: here is how you include JQuery in SS. Rather than having random links to a variety of CDNs. 

I think we also have to be pragmatic.  What do you think?

**Comment by GuySartorelli** [comment]
I think building a whole module just to say "should some random module include javascript?" is going to require a lot of buy-in from many different community members to get their modules using the new thing - which encourages continued use of jQuery when we should be encouraging people not to use it.

Without the buy-in, which would be hard to get, we'd have a whole module or standardised configuration point that's only used by this module anyway.... in which case we're better off just adding it directly to this module.

For now, please just add the configuration option for _this_ module.

**Comment by GuySartorelli** [comment]
> from what I know https:// is better than // - but I can remember exactly why.

Looks like my understanding on that is a bit outdated - just use https.
https://stackoverflow.com/questions/28446314/why-use-protocol-relative-urls-at-all

**Comment by michalkleiner** [comment]
So how does the CMS get its jQuery now? It still needs it, right?

**Comment by michalkleiner** [comment]
For example, [Userforms](https://github.com/silverstripe/silverstripe-userforms/blob/5/code/Control/UserDefinedFormController.php#L81) have a config to block its default JS and if it's not blocked, it includes its own jQuery shipped with the module.

**Comment by GuySartorelli** [comment]
> So how does the CMS get its jQuery now? It still needs it, right?

It's built into the vendor bundle and exposed as a global once that loads in. So it's always available in the CMS, but can't be pulled into the front-end anymore without pulling the rest of the admin dependencies.
It was an oversight at the time to change that, but I don't think we should change it back.

**Comment by xini** [comment]
This is a bit late, but we are using https://github.com/xini/silverstripe-requirements-resolver/ to replace all frontend calls to jquery with one unified version defined in config. Maybe something like this would be an option?

**Comment by GuySartorelli** [comment]
@sunnysideup Still need configuration changes, PR for changelog, and there's a phplinting issue.

**Comment by NightJar** [comment]
I built [this](https://gist.github.com/NightJar/7d3c0053c89f6fbc6c7de9dccfb32d83) as a drop-in 'fix' (workaround) for the admin jquery & entwine disappearance (or relocation).

It would allow continued use of this module without issue (until it is updated).

**Comment by GuySartorelli** [comment]
I've made the requested changes. The changelog PR is no longer necessary since this isn't being released in a minor version anymore. This should at least make the module usable again for CMS 4 projects.

**Review by emteknetnz — APPROVED** [review]

**Comment by emteknetnz** [comment]
Approved with the understanding that we won't merge this up to CMS 5 which doesn't include jquery.  While I don't particularly like getting jquery from a CDN, we need to be pragmatic that so that people who run composer update on CMS 4 don't have existing code suddenly break.

**Pull request merged**